### ISSUE TITLE
Android out of memory crashes are not captured

### ISSUE THREAD

**Issue by troy-lamerton** [bugsnag/bugsnag-unity#531]
### Describe the bug

On Android, when C# code allocates too much memory, the app crashes and this is not reported to Bugsnag.

### Steps to reproduce
1. Call a method that allocates a lot of memory (see snippet below)
2. Wait for the app to crash
3. Launch the app again (and wait a few minutes)
4. Nothing appears on the Bugsnag dashboard

### Environment
* Bugsnag version: `6.1.0`
* Unity version: `2019.4.16f1`
* iOS/Android/macOS/Windows/browser version: Android 11
* simulator/emulator or physical device: physical device
* Initializing bugsnag via a Unity GameObject or in code?: code
* Player Settings:
    * Scripting backend (Mono or IL2CPP): IL2CPP
    * API compatibility level for .NET: 4.x
    * Stack Trace level for all error types (None/ScriptOnly/Full): ScriptOnly

### Example code snippet

Call this method after `Bugsnag.Start` to cause an out of memory crash.

```csharp
public void CrashGameOutOfMemory() {
    var i = 0;
    var arrays = new List<byte[]>();
    while (i < 99999999) {
        const int fiftyMB = 50 * 1024 * 1024;
        arrays.Add(new byte[fiftyMB]);
        i++;
    }
}
```

### Question

Is this something which is supposed to be captured by bugsnag-android? And if not, is there a reason for that or an ETA on when it might be supported?

**Comment by xander-jones** [comment]
Hey @troy-lamerton, I'm seeing an `OutOfMemoryException` raised when running your reproduction example on an Android device using both `bugsnag-unity` v6.1.0 and v6.3.0. I would expect this error to be captured. In my case though using your reprouction, the app did not crash as the memory was trimmed, but the error was reported regardless. If the app did crash because of this, I'd expect the report to be sent upon re-launch of the app 🤔 . What device & Android version are you running this on? Any custom rules to stop the OS trimming memory?

![image](https://user-images.githubusercontent.com/9535852/160118873-433e2b64-6ba9-45f2-b979-c626a0b6cace.png)

Additional quick checks: Do you have any callbacks in your configuration, or discard rules on your Bugsnag project that would otherwise discard this event?

**Comment by troy-lamerton** [comment]
- We don't use callbacks with Bugsnag
- There aren't any discard rules that would discard this event

That is interested you see an `OutOfMemoryException`. When I do this reproduce the app always crashes abruptly, (killed by Android's Low Memory Killer I think) without any obvious errors in logcat. Are you using IL2CPP?

Reproduced on these devices with similar results (abrupt crash and no event in Bugnsag after relaunch):

- Xiaomi Mi 9 (Android 11)
- Samsung Bote 9 (Android 10)

This is the logcat from the app process (for the Mi 9): https://gist.github.com/troy-lamerton/62f707b8a06732dcc998bf7d44e3bc06

In-case it helps, our config looks like this:

```cs
var enabledErrorTypes = new EnabledErrorTypes();
enabledErrorTypes.UnityLog = false; // do not report Debug.LogException, Debug.LogError, etc
var config = new Configuration(bugsnagApiKey) {
    ReleaseStage = "production"
    AppVersion = "64.0.0",
    EnabledErrorTypes = enabledErrorTypes,
    AutoTrackSessions = true,
    EnabledReleaseStages = new [] {
        "production",
        "test"
    },
    PersistUser = true,
    SendLaunchCrashesSynchronously = false,
    LaunchDurationMillis = 0L, // we manually mark launch as completed
};
Bugsnag.Start(config);
```

**Comment by yousif-bugsnag** [comment]
Hi @troy-lamerton, thanks for sharing that config.

Apologies, if your app is being killed by the Low Memory Killer it's actually not the case that we would expect these crashes to be reported - sorry for the confusion there. 

On Android, Bugsnag will automatically capture any unhandled exception occurring in the JVM, including `OutOfMemoryException` errors. However for cases where the Android OS (or the LMK) terminates the app (rather than an `OutOfMemoryException` in the JVM), we don't currently support capture of those terminations. 

Interestingly when I also attempted to reproduce this (using IL2CPP) I also saw an `OutOfMemoryException` reported, and the app didn't crash. I think that whether the JVM raises an `OutOfMemoryException` or the LMK gets there first could be unpredictable.

We've found that OOM terminations by the LMK on Android are difficult to detect reliably with the available Android APIs. For example `REASON_LOW_MEMORY` in the `ApplicationExitReason` API reports to apps that were killed because other apps are consuming memory, which is not very useful. On iOS we use heuristics to determine if an unexpected app termination was due to an OOM event or some other reason, but this isn't always entirely accurate, and so we don't have any plans to implement something similar for Android at this time. 

Improving OOM handling for Android in general is still on our roadmap and is something we'd like to support in the long term, however at this stage I don't have any firm timeframes for when that might be.
